### PR TITLE
Replace EntityRuler with SpanRuler implementation

### PR DESCRIPTION
**Pull request by adrianeboyd**
## Description

Remove `EntityRuler` and rename the `SpanRuler`-based `future_entity_ruler` to `entity_ruler`.

Main changes:

* It is no longer possible to load patterns on init as with `EntityRuler(patterns=)`.
* The older serialization formats (`patterns.jsonl`) are no longer supported and the related tests are removed.
* The config settings are only stored in the config, not in the serialized component (in particular the `phrase_matcher_attr` and overwrite settings).

### Types of change

?

## Checklist

- [x] I confirm that I have the right to submit this contribution under the project's MIT license.
- [x] I ran the tests, and all new and existing tests passed.
- [x] My changes don't require a change to the documentation, or if they do, I've added all required information.

### CODE REVIEW

**Comment by adrianeboyd** [comment]
I'm not really sure how to handle the docs. You could have `/api/entityruler` forward to `/api/spanruler`, but it's hard to cover both configs and options in a non-confusing on the span ruler page.

Instead I removed the `class` tag and source and added the warning at the top.

**Comment by svlandeg** [comment]
> I'm not really sure how to handle the docs. You could have `/api/entityruler` forward to `/api/spanruler`, but it's hard to cover both configs and options in a non-confusing way on the span ruler page.
> 
> Instead I removed the `class` tag and source and added the warning at the top.

I was also going to suggest to just keep both pages separate, highlighting the reduced options on `entityruler` and then refer to `spanruler` to emphasize that that actually has more options/features.

File: website/docs/api/entityruler.md [hunk 1]
```diff
@@ -1,13 +1,19 @@
...
 ---
 
+<Infobox title="New in v4" variant="warning">
+
+As of spaCy v4, there is no separate `EntityRuler` class. The entity ruler is
+implemented as a special case of the `SpanRuler` component. See the
+[`SpanRuler`](/api/spanruler) API docs for the full API.
+
+</Infobox>
+
```
Comment (svlandeg) [inline website/docs/api/entityruler.md:16]: Perhaps also add a note about `remove()` vs `remove_by_id()` ?

**Review by svlandeg — COMMENTED** [review]
This is pretty satisfying :-)

**Comment by adrianeboyd** [comment]
A remaining question is how to deal with references to `EntityRuler` (with the class name) in the docs since the class no longer exists.

**Comment by svlandeg** [comment]
> A remaining question is how to deal with references to `EntityRuler` (with the class name) in the docs since the class no longer exists.

I did a quick search and it looks like pretty much all references on other pages in the docs ( `_architecture.md`, `processing-pipelines.md` etc) can be updated to refer to the new `SpanRuler` / `api/spanruler` instead. Did you have any specific problematic cases in mind?

**Comment by svlandeg** [comment]
Adriane, is there anything I can do to help on this PR?

**Comment by adrianeboyd** [comment]
I think it's confusing to switch the docs entirely to `SpanRuler` in these cases because you do have both `entity_ruler` and `span_ruler` and many people will be expecting to find docs about the entity ruler in general. I also don't think switching to "entity ruler" or `entity_ruler` makes sense because it doesn't line up with the rest of the docs, so in general I'm just still not really sure exactly what to do.

**Pull request merged**